azure-proto-core/Resources/UserAssignedIdentity.cs

```diff
@@ -0,0 +1,129 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+namespace azure_proto_core.Resources
+{
+    using System;
+    using System.Text.Json;
+    using Azure.Core;
+
+    public class UserAssignedIdentity
+    {
+        public Guid? ClientId { get; set; }
+
+        public Guid? PrincipalId { get; set; }
+
+        public UserAssignedIdentity() { }
+
+        public UserAssignedIdentity(Guid clientId, Guid principalId)
+        {
+            ClientId = clientId;
+            PrincipalId = principalId;
+        }
+
+        public int CompareTo(UserAssignedIdentity other)
+        {
+            if (other == null)
+                return 1;
+
+            if (!ClientId.HasValue && !other.ClientId.HasValue)
+                return 0;
+
+            if (!ClientId.HasValue)
+                return -1;
+
+            if (!other.ClientId.HasValue)
+                return 1;
+
+            if (ClientId.Value.CompareTo(other.ClientId.Value) == 1 && PrincipalId.Value.CompareTo(other.PrincipalId.Value) == 1)
```

Comment:
I think this is a little difficult to understand logically.  I would prefer something like:

``` c#
var clientCompare = ClientId.Value.CompareTo(other.ClientId.Value);
if (clientCompare == 0)
{
    return PrincipalId.Value.CompareTo(other.PrincipalId.Value);
}

return clientCompare;
```

I think this is easier to inspect and determine that it is correct (if ClientIds are equal, only then does the principalId comparison make sense)